### ISSUE TITLE
different implementations of mackinnoncrit in coint vs adfulller

### ISSUE THREAD

**Issue by tsando** [statsmodels/statsmodels#3967]
Can someone please explain why the `mackinnoncrit ` implementation is different in `coint` vs `adfuller`? In `adfuller` we have this as

`critvalues = mackinnoncrit(N=1, regression=regression, nobs=nobs)`

whereas in `coint` we have this as:

`mackinnoncrit(N=k_vars, regression=trend, nobs=nobs - 1)`

which results in a different set of critical values for the same input. I would actually expect these to be the same, since they are supposed to be used for similar purposes?

As an example of values I see please see screenshot below.

![image](https://user-images.githubusercontent.com/13929456/31129220-cd99acc8-a84b-11e7-9324-603860fe705d.png)

**Comment by josef-pkt** [comment]
Based on what I remember, the asymptotic distribution depends on the number of series that are in the cointegrating relationship. i.e. the `N=k_vars` argument.

**Comment by tsando** [comment]
yes, but the point is that the above should be equivalent. If you look at the definition of `k_vars` (line 1003):

```
    nobs, k_vars = y1.shape
    k_vars += 1   # add 1 for y0
```

this seems to be adding 1 to `k_vars`, instead of leaving it as 1. In `adfuller`, this is set always to `N=1`. Given that the `coint` function is meant to return the adf critical values, this should give the same values as running the equivalent `adfuller`, which as shown by my screenshot, is not the case...

**Comment by josef-pkt** [comment]
I will look at the details again later today.

My impression at the moment is that when you do adfuller(x=r.resid), then adfuller does not take into account that resid is from a cointegrating regression and the adfuller p-values are not the correct ones.
The correct distribution has two integrated, N=2, variables.

(It hass been a while since I looked at this, so I need to check again.)

**Comment by tsando** [comment]
ok, thanks, yes, would be great if you could check this difference in N... btw, please see my updated comment above, although I think you addressed it already. However we do need to understand which one is correct, or if a fix is required

**Comment by tsando** [comment]
btw, this is in `coint` lines 1027-1029:

```
        crit = mackinnoncrit(N=k_vars, regression=trend, nobs=nobs - 1)
        #  nobs - 1, the -1 is to match egranger in Stata, I don't know why.
        #  TODO: check nobs or df = nobs - k
```
so, based on the commented lines, it seems the author wasn't sure about the inputs of `mackinnoncrit`...

**Comment by tsando** [comment]
Also in the `mackinnoncrit` def, you see this

```
    Parameters
    ----------
    N : int
        The number of series of I(1) series for which the null of
        non-cointegration is being tested.  For N > 12, the critical values
        are linearly interpolated (not yet implemented).  For the ADF test,
        N = 1.
```

So this seems to suggest that for the ADF test you should always use 1 and since we are doing this in `coint` this implies `k_var` should be always 1 right?

**Comment by josef-pkt** [comment]
about the previous comment, nobs versus nobs-1 versus nobs-k:
That is different. This is just a small sample degrees of freedom correction which is irrelevant in large sample, large nobs. (i.e. asymptotically irrelevant) 
In contrast, the number of variables itself affects the distribution independent of the number of observations, even asymptotically.

(I did all the rewriting of coint because it was buggy before.)

**Comment by josef-pkt** [comment]
ADF and coint engle granger are NOT the same. 
ADF is a unit root test
coint is a cointegration test

That coint used adf is "coincidental". Even though they produce the same test statistic, the distribution of the test statistics is different under the two settings.
i.e. 
adf **assumes** that we have one integrated series, 
coint **assumes** that we have N (or k) cointegrated series.

**Comment by josef-pkt** [comment]
as reference: See the description N=1 (ADF) versus N > 1 (EG) in 
http://qed.econ.queensu.ca/working_papers/papers/qed_wp_1227.pdf

**Comment by tsando** [comment]
@josef-pkt thanks for clarifying this - that paper is indeed very useful to understand the difference... i fear however, a few users might use adf to test stationarity for two cointegrated series, without realising the critical values are not correct... do you think this should be made clearer in the documentation?

**Comment by josef-pkt** [comment]
> do you think this should be made clearer in the documentation?

Yes, that's a good idea.
Do you want to write something for the Notes sections of the functions?

(In general we don't provide much extra information in the documentation, but that's mainly because we don't have the wo/man power, and because it's often not clear what we can expect that users of specific functions know.)

**Comment by josef-pkt** [comment]
I added the FAQ label, so we can leave the issue open.

**Comment by ymzz277** [comment]
two more observations on this:
1. in 0.6.0 and 0.8.0, coint gave different pvalues for 'c'
2. in 0.8.0, in adfuller step, regression is always 'nc', while mackinnonp step takes in trend. In my case, it is 'c'. The resulting pvalue changes dramatically depending on whether 'c' or 'nc' is passed in. 
Please clarify if the mackinnonp step should take a fixed 'nc' as well

**Comment by josef-pkt** [comment]
@ymzz277  (quickly written answers)
to 1. In 0.6 and before, coint was hidden (not in documentation) and had bugs

to 2. As far as I remember this should be correct. I thought we had a question/issue before but don't have time to find it right now.
The trend is included in the cointegrating regression, but not in the adf statistic

```
     trend : str {'c', 'ct'}
         trend term included in regression for cointegrating equation
```
As far as I remember this matches the implementation used as reference/comparison cases.
(I'm not sure about different possibilities in coint engle granger.
For VECM we looked more carefully at the difference in the meaning of trend in the difference equation versus trend in the cointegrating equation. But it's more than I year ago that I looked at it.)

**Comment by josef-pkt** [comment]
There is currently a bug in the autolag option #3966  (which wasn't linked to this issue)

**Comment by ymzz277** [comment]
I read http://qed.econ.queensu.ca/working_papers/papers/qed_wp_1227.pdf
Although it is a bit complicated, I still think the parameters in the two calls should be consistent:
1. res_adf = adfuller(res_co.resid, maxlag=maxlag, autolag=None,  regression='nc')
2. crit = mackinnoncrit(N=k_vars, regression=trend, nobs=nobs - 1)
Depending on the trend value, the current implementation might use inconsistent values in the two calls, which makes no sense to me.

**Comment by josef-pkt** [comment]
According to the unit tests, AFAICS, our numbers agree with the Stata user function `egranger` for trend='c', 'ct' and 'ctt'. AFAIR, I didn't find any other package to compare to for engle granger.

If the trend is correctly modeled in the regression for the cointegrating equation, then there should not be any trend left over in the residuals.

see after equ. (3)
"Because there is a constant term in (1), there is no need to include one in (2) or (3)."
the same will apply for trend
OLS  has by construction residuals with zero mean. The only difference can come from having to drop different number of initial observations, as they mention.

Also having a non-zero constant in the difference equation would imply a trend in the level/cointegrating equation. (That is used in VECM to include both.)

**Comment by ymzz277** [comment]
I fully agree with your quotes after equ (3)
I guess my confusion is with 
"If N = 1, this procedure is equivalent to one variant of the ordinary DF test (see
below), in which one runs the regression
∆zt = α1 + γzt−1 + εt
and tests for γ = 0."
Does it mean when N!=1, there would be no α1? Also for cointegration testing, shouldn't N>=1 always?

**Comment by josef-pkt** [comment]
**co**integration requires N>=2, and then we use a cointegrating regression in engle granger coint
If N=1, then we only have one variable which can be **integrated** but without the **co**.

> This procedure is equivalent

For N=1, i.e. unit root testing, we could use a preliminary regression on the trend to remove the trend before doing adf with "nc". We can `absorb` variables in OLS by taking it out of each of the involved variables by using residuals of preliminary regressions. In the diff/lag case, there will again be a small difference because of the initial observations, but the two procedures become identical in large samples. (Although, I have never seen or tried this.)

(everything is of course AFAIU)

**Comment by ymzz277** [comment]
I have tried it and it is the case. 
The issue is if I care about N>=2, the z has no drift due to the regression on y with constant ('c'), why the pvalue needs to be determined by regression='c' in the 2nd step?

**Comment by josef-pkt** [comment]
The difference is between computing the test statistic and the distribution of the test statistic.

For computing the test statistic there are equivalent procedures. 
However, the asymptotic distribution and pvalue depends on the overall model, i.e. we have to take the 'c' or trend from the first step, cointegrating equation regression into account even though we don't use it in the second step adf computation.

**Comment by ymzz277** [comment]
Thank you
